### PR TITLE
[jvm-packages] Exposed baseMargin

### PR DESCRIPTION
**Pull request by superbobry**
The users can now initialize boosting with predictions from an external model. This feature is implemented via RDD API and therefore does not require the user to know the partitioning of the training set.

### CODE REVIEW

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 1]
```diff
@@ -130,6 +130,9 @@ object XGBoost extends Serializable {
               trainingSet.setGroup(xgBoostConfMap("groupData").asInstanceOf[Seq[Seq[Int]]](
                   TaskContext.getPartitionId()).toArray)
             }
+            if (params.contains("base_margin")) {
+              trainingSet.setBaseMargin(params("base_margin").asInstanceOf[Array[Float]])
```
Comment (superbobry) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: I can rewrite it in a way that it also supports `Array[Array[Float]]` and fails if the parameter is of different type. What do you think?

**Comment by superbobry** [comment]
I start to think that extending params might not be the best option, because the resulting parameter would be weakly typed and would required documentation to be discoverable. Stay tuned.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 2]
```diff
@@ -81,8 +79,13 @@ object XGBoost extends Serializable {
...
       rabitEnv: java.util.Map[String, String],
-      numWorkers: Int, round: Int, obj: ObjectiveTrait, eval: EvalTrait,
-      useExternalMemory: Boolean, missing: Float = Float.NaN): RDD[Booster] = {
+      numWorkers: Int,
+      round: Int,
+      obj: ObjectiveTrait,
+      eval: EvalTrait,
+      useExternalMemory: Boolean,
+      missing: Float,
+      baseMargin: RDD[Float]): RDD[Booster] = {
```
Comment (superbobry) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: I've seen the discussion in #2223 and decided not to modify LabeledPoint at the moment. This could be done in a separate PR for both `baseMargin` and `groupData`.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 3]
```diff
@@ -53,97 +51,92 @@ case class TrackerConf(workerConnectionTimeout: Long, trackerImpl: String)
...
-          Rabit.shutdown()
-          throw new XGBoostError(s"detect the empty partition in training dataset, partition ID:" +
-            s" ${TaskContext.getPartitionId().toString}")
+    partitionedTrainingSet.zipPartitions(partitionedBaseMargin) { (trainingSamples, baseMargin) =>
+      val cacheFileName = if (useExternalMemory) {
+        s"$appName-${TaskContext.get().stageId()}-" +
+            s"dtrain_cache-${TaskContext.getPartitionId()}"
+      } else {
+        null
+      }
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: we should put this after emptyPartition check

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 4]
```diff
@@ -53,97 +51,92 @@ case class TrackerConf(workerConnectionTimeout: Long, trackerImpl: String)
...
+      Rabit.init(rabitEnv)
+
+      val partitionItr = fromDenseToSparseLabeledPoints(trainingSamples, missing)
+      val trainingMatrix = new DMatrix(new JDMatrix(partitionItr, cacheFileName))
+      try {
+        if (params.contains("groupData") && params("groupData") != null) {
+          trainingMatrix.setGroup(params("groupData").asInstanceOf[Seq[Seq[Int]]](
+            TaskContext.getPartitionId()).toArray)
+        }
+
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: for these empty lines, I'm not sure if they are really necessary....

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala [hunk 5]
```diff
@@ -152,7 +151,7 @@ abstract class XGBoostModel(protected var _booster: Booster)
    * @param testSet test set represented as RDD
    * @param missingValue the specified value to represent the missing value
    */
-  def predict(testSet: RDD[MLDenseVector], missingValue: Float): RDD[Array[Array[Float]]] = {
+  def predict(testSet: RDD[MLDenseVector], missingValue: Float): RDD[Array[Float]] = {
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala]: let's take keep the original return type deprecate it and build a new API returning the the preferred type, and we'd better do it in a separate PR

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/DataBatch.java [hunk 6]
```diff
@@ -13,101 +14,83 @@
...
-            ret.featureIndex[offset + j] = j;
+      for (int i = 0; i < batch.size(); i++) {
+        LabeledPoint labeledPoint = batch.get(i);
+        rowOffset[i] = offset;
+        label[i] = labeledPoint.label;
+        if (labeledPoint.indices != null) {
+          System.arraycopy(
+                  labeledPoint.indices, 0,
+                  featureIndex, offset,
+                  labeledPoint.indices.length);
```
Comment (CodingCat) [inline jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/DataBatch.java]: for this simple expression, we do not need to make it multi-lines

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala (hunk 3 above)
Comment (superbobry) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Done.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala (hunk 4 above)
Comment (superbobry) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Removed those.

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/DataBatch.java (hunk 6 above)
Comment (superbobry) [inline jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/DataBatch.java]: Reverted.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala (hunk 5 above)
Comment (superbobry) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala]: Reverted. Do you have a preference regarding the naming/location of the new API?

Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala]: how about `predictWithFlatten(testSet: RDD[MLDenseVector], missing: Float)`?

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala [hunk 7]
```diff
@@ -334,6 +334,13 @@ abstract class XGBoostModel(protected var _booster: Booster)
 }
 
 object XGBoostModel extends MLReadable[XGBoostModel] {
+  private[spark] def apply(booster: Booster, isClassification: Boolean): XGBoostModel = {
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala]: do we call this in anywhere?

File: jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/DMatrix.scala [hunk 8]
```diff
@@ -129,11 +129,11 @@ class DMatrix private[scala](private[scala] val jDMatrix: JDMatrix) {
...
-   * if specified, xgboost will start from this init margin
-   * can be used to specify initial prediction to boost from
-   *
-   * @param baseMargin base margin
-   */
+    * if specified, xgboost will start from this init margin
+    * can be used to specify initial prediction to boost from
+    *
+    * @param baseMargin base margin
+    */
```
Comment (CodingCat) [inline jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/DMatrix.scala]: I do not think the indention follows the scala style guide

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 9]
```diff
@@ -53,97 +51,90 @@ case class TrackerConf(workerConnectionTimeout: Long, trackerImpl: String)
...
+          if (dFeatures.values(i) != missing) {
             indices += i
-            values += dVector.values(i)
+            values += dFeatures.values(i)
           }
         }
-        val sparseVector = new SparseVector(dVector.values.length, indices.toArray,
-          values.toArray)
-        sparseLabeledPoints += MLLabeledPoint(labelPoint.label, sparseVector)
+
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: extra empty line

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 10]
```diff
@@ -53,97 +51,90 @@ case class TrackerConf(workerConnectionTimeout: Long, trackerImpl: String)
...
     import DataUtils._
-    val partitionedTrainingSet = repartitionData(trainingSet, numWorkers)
+
+    val partitionedTrainingSet = if (trainingSet.getNumPartitions != numWorkers) {
+      logger.info(s"repartitioning training set to $numWorkers partitions")
+      trainingSet.repartition(numWorkers)
+    } else {
+      trainingSet
+    }
+
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: extra empty line

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 11]
```diff
@@ -53,97 +51,90 @@ case class TrackerConf(workerConnectionTimeout: Long, trackerImpl: String)
...
+    val partitionedBaseMargin = baseMargin.repartition(partitionedTrainingSet.getNumPartitions)
+
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: extra empty line

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 12]
```diff
@@ -53,97 +51,90 @@ case class TrackerConf(workerConnectionTimeout: Long, trackerImpl: String)
...
-          Rabit.shutdown()
-          throw new XGBoostError(s"detect the empty partition in training dataset, partition ID:" +
-            s" ${TaskContext.getPartitionId().toString}")
+    partitionedTrainingSet.zipPartitions(partitionedBaseMargin) { (trainingSamples, baseMargin) =>
+      if (trainingSamples.isEmpty) {
+        throw new XGBoostError(
+          s"detected an empty partition in the training data, partition ID:" +
+              s" ${TaskContext.getPartitionId()}")
+      }
+
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: extra empty line

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 13]
```diff
@@ -53,97 +51,90 @@ case class TrackerConf(workerConnectionTimeout: Long, trackerImpl: String)
...
+      val cacheFileName = if (useExternalMemory) {
+        s"$appName-${TaskContext.get().stageId()}-" +
+            s"dtrain_cache-${TaskContext.getPartitionId()}"
+      } else {
+        null
+      }
+
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: extra empty line

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 14]
```diff
@@ -53,97 +51,90 @@ case class TrackerConf(workerConnectionTimeout: Long, trackerImpl: String)
...
+      rabitEnv.put("DMLC_TASK_ID", TaskContext.getPartitionId().toString)
+      Rabit.init(rabitEnv)
+
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: extra empty line

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala (hunk 5 above)
Comment (superbobry) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala]: I thought about this more and I am not sure it is worth going through the deprecation cycle for this particular API, because

1. it does not satisfy the user expectations regarding the shape of the output;
2. the users are probably doing the same hack as the tests: `.collect()(i)`, which is easy to adjust to the new API;
3. there're no official releases of xgboost4j, no published JARs, nothing, therefore the deprecation cycle is hard to implement properly;
4. any other name than `predict` would be harder to guess and it's a shame to have it taken by the broken implementation.

What do you think?

File: jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/DMatrix.scala (hunk 8 above)
Comment (superbobry) [inline jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/DMatrix.scala]: Yep, sorry, it's IntelliJ pushing ScalaDoc formatting on me.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala (hunk 9 above)
Comment (superbobry) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Removed.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala (hunk 10 above)
Comment (superbobry) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Removed.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala (hunk 11 above)
Comment (superbobry) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Removed.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala (hunk 12 above)
Comment (superbobry) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Removed.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala (hunk 14 above)
Comment (superbobry) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Removed.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala (hunk 13 above)
Comment (superbobry) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Removed.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala (hunk 5 above)
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala]: ok, let's move forward with your approach

Comment (superbobry) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala]: Cool! I'll submit it in a dedicated PR.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala (hunk 7 above)
Comment (superbobry) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala]: Yep, in `XGBoost.postTrackerReturnProcessing`.

**Comment by superbobry** [comment]
I think this is good to merge. I'd submit `predict` API change in the followup PR.

**Comment by CodingCat** [comment]
merged, thanks
